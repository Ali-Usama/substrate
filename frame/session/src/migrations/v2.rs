// This file is part of Substrate.

// Copyright (C) Parity Technologies (UK) Ltd.
// SPDX-License-Identifier: Apache-2.0

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// 	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::{Pallet, Config, LOG_TARGET};
use frame_support::{
    traits::{Get, OnRuntimeUpgrade},
    dispatch::GetStorageVersion,
    storage::storage_prefix,
    weights::Weight,
};
use sp_io::{storage::clear_prefix, KillStorageResult};

pub struct MigrateToV2<T>(sp_std::marker::PhantomData<T>);
impl<T: Config> OnRuntimeUpgrade for MigrateToV2<T> {
    #[cfg(feature = "try-runtime")]
    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
        let onchain = Pallet::<T>::on_chain_storage_version();
        ensure!(onchain < 2, "pallet_session::MigrateToV2 migration can be deleted");
        Ok(Vec::new())
    }

    fn on_runtime_upgrade() -> Weight {
        let current = Pallet::<T>::current_storage_version();
        let onchain = Pallet::<T>::on_chain_storage_version();

        if onchain > 1 {
            log::info!(target: LOG_TARGET, "pallet_session::MigrateToV2 should be removed");
            return T::DbWeight::get().reads(1)
        }

        let prefix = storage_prefix(b"Session", b"StoredRange");
        let keys_removed_stored_range = match clear_prefix(&prefix, None) {
			KillStorageResult::AllRemoved(value) => value,
			KillStorageResult::SomeRemaining(value) => {
				log::error!(
					"`clear_prefix` failed to remove all keys. THIS SHOULD NEVER HAPPEN! ðŸš¨",
				);
				value
			},
		} as u64;

        let prefix = storage_prefix(b"Session", b"HistoricalSessions");
        let keys_removed_historical_sessions = match clear_prefix(&prefix, None) {
			KillStorageResult::AllRemoved(value) => value,
			KillStorageResult::SomeRemaining(value) => {
				log::error!(
					"`clear_prefix` failed to remove all keys. THIS SHOULD NEVER HAPPEN! ðŸš¨",
				);
				value
			},
		} as u64;

        let keys_removed = keys_removed_stored_range + keys_removed_historical_sessions;
		log::info!("Removed {} keys ðŸ§¹", keys_removed);

        current.put::<Pallet<T>>();

        T::DbWeight::get().reads_writes(keys_removed, keys_removed)
    }

    #[cfg(feature = "try-runtime")]
    fn post_upgrade(_state: Vec<u8>) -> Result<(), &'static str> {
        let onchain = Pallet::<T>::on_chain_storage_version();
        ensure!(onchain == 2, "pallet_session::MigrateToV2 needs to be run");
        Ok(())
    }
}

#[cfg(test)]
mod test {
	use super::*;
    use crate::mock::{new_test_ext, Test as T};
    use frame_support::{weights::RuntimeDbWeight, Twox64Concat};

    #[test]
	fn migration_to_v1_works() {
        #[frame_support::storage_alias]
        type HistoricalSessions = StorageMap<Session, Twox64Concat, u64, u64>;

        #[frame_support::storage_alias]
        type StoredRange = StorageValue<Session, u64>;

		let mut ext = new_test_ext();

		ext.execute_with(|| {
			HistoricalSessions::insert(1, 10);
            StoredRange::set(Some(5));

			assert!(HistoricalSessions::iter_keys().collect::<Vec<_>>().len() > 0);
            assert_eq!(StoredRange::get(), Some(5));
		});

		ext.commit_all().unwrap();

		ext.execute_with(|| {
            let weight: RuntimeDbWeight = <T as frame_system::Config>::DbWeight::get();

			assert_eq!(
				MigrateToV2::<T>::on_runtime_upgrade(),
				weight.reads_writes(2, 2),
			);

			assert!(HistoricalSessions::iter_keys().collect::<Vec<_>>().len() == 0);
            assert_eq!(StoredRange::get(), None);
		})
	}
}