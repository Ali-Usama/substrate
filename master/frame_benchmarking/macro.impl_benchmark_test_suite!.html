<!DOCTYPE html>
<html lang="en">
<head><link rel="canonical" href="https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/macro.impl_benchmark_test_suite!.html"><meta http-equiv="refresh" content="0;URL='https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/macro.impl_benchmark_test_suite!.html'">
    <meta http-equiv="refresh" content="0;URL=macro.impl_benchmark_test_suite.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.impl_benchmark_test_suite.html">macro.impl_benchmark_test_suite.html</a>...</p>
    <script>location.replace("macro.impl_benchmark_test_suite.html" + location.search + location.hash);</script>
</body>
</html>