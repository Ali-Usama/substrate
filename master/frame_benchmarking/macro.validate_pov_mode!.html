<!DOCTYPE html>
<html lang="en">
<head><link rel="canonical" href="https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/macro.validate_pov_mode!.html"><meta http-equiv="refresh" content="0;URL='https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/macro.validate_pov_mode!.html'">
    <meta http-equiv="refresh" content="0;URL=macro.validate_pov_mode.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.validate_pov_mode.html">macro.validate_pov_mode.html</a>...</p>
    <script>location.replace("macro.validate_pov_mode.html" + location.search + location.hash);</script>
</body>
</html>