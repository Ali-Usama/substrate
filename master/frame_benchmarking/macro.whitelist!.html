<!DOCTYPE html>
<html lang="en">
<head><link rel="canonical" href="https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/macro.whitelist!.html"><meta http-equiv="refresh" content="0;URL='https://paritytech.github.io/polkadot-sdk/master/frame_benchmarking/macro.whitelist!.html'">
    <meta http-equiv="refresh" content="0;URL=macro.whitelist.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.whitelist.html">macro.whitelist.html</a>...</p>
    <script>location.replace("macro.whitelist.html" + location.search + location.hash);</script>
</body>
</html>