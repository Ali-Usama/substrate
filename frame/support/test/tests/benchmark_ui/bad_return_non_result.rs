use frame_benchmarking::v2::*;
#[allow(unused_imports)]
use frame_support_test::Config;

#[benchmarks]
mod benchmarks {
	#[benchmark]
	fn bench() -> Option<String> {
		#[block]
		{}
	}
}

fn main() {}
