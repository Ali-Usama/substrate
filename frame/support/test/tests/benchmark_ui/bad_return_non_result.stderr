error: expected `Result`
 --> tests/benchmark_ui/bad_return_non_result.rs:8:16
  |
8 |     fn bench() -> Option<String> {
  |                   ^^^^^^
